File: clients/ios/AxiomVault/Sources/Core/VaultCore.swift
import Foundation

/// Errors that can occur during vault operations
enum VaultError: Error, LocalizedError {
    case initializationFailed
    case creationFailed(String)
    case openFailed(String)
    case operationFailed(String)
    case invalidHandle
    case invalidPath
    case jsonParsingFailed

    var errorDescription: String? {
        switch self {
        case .initializationFailed:
            return "Failed to initialize AxiomVault"
        case .creationFailed(let msg):
            return "Failed to create vault: \(msg)"
        case .openFailed(let msg):
            return "Failed to open vault: \(msg)"
        case .operationFailed(let msg):
            return "Vault operation failed: \(msg)"
        case .invalidHandle:
            return "Invalid vault handle"
        case .invalidPath:
            return "Invalid path"
        case .jsonParsingFailed:
            return "Failed to parse JSON response"
        }
    }
}

/// Information about a vault
struct VaultInfo {
    let vaultId: String
    let rootPath: String
    let fileCount: Int
    let totalSize: Int64
    let version: Int
}

/// File or directory entry in the vault
struct VaultEntry: Identifiable, Codable {
    let id: UUID
    let name: String
    let isDirectory: Bool
    let size: Int64?

    enum CodingKeys: String, CodingKey {
        case name
        case isDirectory = "is_directory"
        case size
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.id = UUID()
        self.name = try container.decode(String.self, forKey: .name)
        self.isDirectory = try container.decode(Bool.self, forKey: .isDirectory)
        self.size = try container.decodeIfPresent(Int64.self, forKey: .size)
    }

    init(name: String, isDirectory: Bool, size: Int64?) {
        self.id = UUID()
        self.name = name
        self.isDirectory = isDirectory
        self.size = size
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(name, forKey: .name)
        try container.encode(isDirectory, forKey: .isDirectory)
        try container.encodeIfPresent(size, forKey: .size)
    }
}

/// Swift wrapper for AxiomVault Rust core
class VaultCore {
    /// Singleton instance
    static let shared = VaultCore()

    /// Whether the core has been initialized
    private var initialized = false

    /// Current vault handle
    private var handle: OpaquePointer?

    /// Lock for thread safety
    private let lock = NSLock()

    private init() {}

    deinit {
        closeVault()
    }

    /// Initialize the FFI layer
    func initialize() throws {
        lock.lock()
        defer { lock.unlock() }

        guard !initialized else { return }

        let result = axiom_init()
        guard result == 0 else {
            throw VaultError.initializationFailed
        }

        initialized = true
    }

    /// Get the library version
    func version() -> String {
        guard let versionPtr = axiom_version() else {
            return "unknown"
        }
        return String(cString: versionPtr)
    }

    /// Create a new vault
    func createVault(at path: String, password: String) throws {
        lock.lock()
        defer { lock.unlock() }

        guard initialized else {
            throw VaultError.initializationFailed
        }

        // Close existing vault if open
        if let existingHandle = handle {
            axiom_vault_close(existingHandle)
            handle = nil
        }

        let pathCStr = path.cString(using: .utf8)!
        let passwordCStr = password.cString(using: .utf8)!

        guard let newHandle = axiom_vault_create(pathCStr, passwordCStr) else {
            let error = getLastError()
            throw VaultError.creationFailed(error)
        }

        handle = newHandle
    }

    /// Open an existing vault
    func openVault(at path: String, password: String) throws {
        lock.lock()
        defer { lock.unlock() }

        guard initialized else {
            throw VaultError.initializationFailed
        }

        // Close existing vault if open
        if let existingHandle = handle {
            axiom_vault_close(existingHandle)
            handle = nil
        }

        let pathCStr = path.cString(using: .utf8)!
        let passwordCStr = password.cString(using: .utf8)!

        guard let newHandle = axiom_vault_open(pathCStr, passwordCStr) else {
            let error = getLastError()
            throw VaultError.openFailed(error)
        }

        handle = newHandle
    }

    /// Close the current vault
    func closeVault() {
        lock.lock()
        defer { lock.unlock() }

        if let currentHandle = handle {
            _ = axiom_vault_close(currentHandle)
            handle = nil
        }
    }

    /// Get vault information
    func getVaultInfo() throws -> VaultInfo {
        lock.lock()
        defer { lock.unlock() }

        guard let currentHandle = handle else {
            throw VaultError.invalidHandle
        }

        guard let infoPtr = axiom_vault_info(currentHandle) else {
            let error = getLastError()
            throw VaultError.operationFailed(error)
        }

        defer {
            axiom_vault_info_free(infoPtr)
        }

        let info = infoPtr.pointee

        let vaultId = info.vault_id != nil ? String(cString: info.vault_id) : ""
        let rootPath = info.root_path != nil ? String(cString: info.root_path) : ""

        return VaultInfo(
            vaultId: vaultId,
            rootPath: rootPath,
            fileCount: Int(info.file_count),
            totalSize: info.total_size,
            version: Int(info.version)
        )
    }

    /// List contents of a directory in the vault
    func listDirectory(at path: String = "/") throws -> [VaultEntry] {
        lock.lock()
        defer { lock.unlock() }

        guard let currentHandle = handle else {
            throw VaultError.invalidHandle
        }

        let pathCStr = path.cString(using: .utf8)!

        guard let jsonPtr = axiom_vault_list(currentHandle, pathCStr) else {
            let error = getLastError()
            throw VaultError.operationFailed(error)
        }

        defer {
            axiom_string_free(jsonPtr)
        }

        let jsonString = String(cString: jsonPtr)
        guard let jsonData = jsonString.data(using: .utf8) else {
            throw VaultError.jsonParsingFailed
        }

        let decoder = JSONDecoder()
        return try decoder.decode([VaultEntry].self, from: jsonData)
    }

    /// Add a file to the vault
    func addFile(from localPath: String, to vaultPath: String) throws {
        lock.lock()
        defer { lock.unlock() }

        guard let currentHandle = handle else {
            throw VaultError.invalidHandle
        }

        let localCStr = localPath.cString(using: .utf8)!
        let vaultCStr = vaultPath.cString(using: .utf8)!

        let result = axiom_vault_add_file(currentHandle, localCStr, vaultCStr)
        if result != 0 {
            let error = getLastError()
            throw VaultError.operationFailed(error)
        }
    }

    /// Extract a file from the vault
    func extractFile(from vaultPath: String, to localPath: String) throws {
        lock.lock()
        defer { lock.unlock() }

        guard let currentHandle = handle else {
            throw VaultError.invalidHandle
        }

        let vaultCStr = vaultPath.cString(using: .utf8)!
        let localCStr = localPath.cString(using: .utf8)!

        let result = axiom_vault_extract_file(currentHandle, vaultCStr, localCStr)
        if result != 0 {
            let error = getLastError()
            throw VaultError.operationFailed(error)
        }
    }

    /// Create a directory in the vault
    func createDirectory(at vaultPath: String) throws {
        lock.lock()
        defer { lock.unlock() }

        guard let currentHandle = handle else {
            throw VaultError.invalidHandle
        }

        let pathCStr = vaultPath.cString(using: .utf8)!

        let result = axiom_vault_mkdir(currentHandle, pathCStr)
        if result != 0 {
            let error = getLastError()
            throw VaultError.operationFailed(error)
        }
    }

    /// Remove a file or directory from the vault
    func removeEntry(at vaultPath: String) throws {
        lock.lock()
        defer { lock.unlock() }

        guard let currentHandle = handle else {
            throw VaultError.invalidHandle
        }

        let pathCStr = vaultPath.cString(using: .utf8)!

        let result = axiom_vault_remove(currentHandle, pathCStr)
        if result != 0 {
            let error = getLastError()
            throw VaultError.operationFailed(error)
        }
    }

    /// Change the vault password
    func changePassword(old oldPassword: String, new newPassword: String) throws {
        lock.lock()
        defer { lock.unlock() }

        guard let currentHandle = handle else {
            throw VaultError.invalidHandle
        }

        let oldCStr = oldPassword.cString(using: .utf8)!
        let newCStr = newPassword.cString(using: .utf8)!

        let result = axiom_vault_change_password(currentHandle, oldCStr, newCStr)
        if result != 0 {
            let error = getLastError()
            throw VaultError.operationFailed(error)
        }
    }

    /// Check if a vault is currently open
    var isVaultOpen: Bool {
        lock.lock()
        defer { lock.unlock() }
        return handle != nil
    }

    // MARK: - Private helpers

    private func getLastError() -> String {
        guard let errorPtr = axiom_last_error() else {
            return "Unknown error"
        }
        defer {
            axiom_string_free(errorPtr)
        }
        return String(cString: errorPtr)
    }
}

File: clients/android/app/src/main/kotlin/com/axiomvault/android/core/VaultCore.kt
package com.axiomvault.android.core

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import java.util.concurrent.atomic.AtomicBoolean
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Errors that can occur during vault operations
 */
sealed class VaultError : Exception() {
    data object InitializationFailed : VaultError() {
        private fun readResolve(): Any = InitializationFailed
        override val message = "Failed to initialize AxiomVault"
    }

    data class CreationFailed(override val message: String) : VaultError()
    data class OpenFailed(override val message: String) : VaultError()
    data class OperationFailed(override val message: String) : VaultError()

    data object InvalidHandle : VaultError() {
        private fun readResolve(): Any = InvalidHandle
        override val message = "Invalid vault handle"
    }

    data object InvalidPath : VaultError() {
        private fun readResolve(): Any = InvalidPath
        override val message = "Invalid path"
    }

    data object JsonParsingFailed : VaultError() {
        private fun readResolve(): Any = JsonParsingFailed
        override val message = "Failed to parse JSON response"
    }
}

/**
 * Information about a vault
 */
data class VaultInfo(
    val vaultId: String,
    val rootPath: String,
    val fileCount: Int,
    val totalSize: Long,
    val version: Int
)

/**
 * File or directory entry in the vault
 */
data class VaultEntry(
    val name: String,
    val isDirectory: Boolean,
    val size: Long?
) {
    val id: String = java.util.UUID.randomUUID().toString()
}

/**
 * JSON model for parsing vault entries from native code
 */
private data class VaultEntryJson(
    val name: String,
    val is_directory: Boolean,
    val size: Long?
)

/**
 * Kotlin wrapper for AxiomVault Rust core via JNI
 */
@Singleton
class VaultCore @Inject constructor() {

    companion object {
        private const val LIBRARY_NAME = "axiom_vault"

        init {
            try {
                System.loadLibrary(LIBRARY_NAME)
            } catch (e: UnsatisfiedLinkError) {
                throw RuntimeException("Failed to load native library: $LIBRARY_NAME", e)
            }
        }
    }

    private val initialized = AtomicBoolean(false)
    private var handle: Long = 0L
    private val mutex = Mutex()
    private val gson = Gson()

    // Native methods (JNI bindings to Rust FFI)
    private external fun nativeInit(): Int
    private external fun nativeVersion(): String
    private external fun nativeCreateVault(path: String, password: String): Long
    private external fun nativeOpenVault(path: String, password: String): Long
    private external fun nativeCloseVault(handle: Long): Int
    private external fun nativeVaultInfo(handle: Long): String?
    private external fun nativeVaultList(handle: Long, path: String): String?
    private external fun nativeAddFile(handle: Long, localPath: String, vaultPath: String): Int
    private external fun nativeExtractFile(handle: Long, vaultPath: String, localPath: String): Int
    private external fun nativeMkdir(handle: Long, vaultPath: String): Int
    private external fun nativeRemove(handle: Long, vaultPath: String): Int
    private external fun nativeChangePassword(handle: Long, oldPassword: String, newPassword: String): Int
    private external fun nativeLastError(): String?

    /**
     * Initialize the FFI layer
     */
    suspend fun initialize() = withContext(Dispatchers.IO) {
        mutex.withLock {
            if (initialized.get()) return@withLock

            val result = nativeInit()
            if (result != 0) {
                throw VaultError.InitializationFailed
            }

            initialized.set(true)
        }
    }

    /**
     * Get the library version
     */
    fun version(): String {
        return try {
            nativeVersion()
        } catch (e: Exception) {
            "unknown"
        }
    }

    /**
     * Create a new vault
     */
    suspend fun createVault(path: String, password: String) = withContext(Dispatchers.IO) {
        mutex.withLock {
            if (!initialized.get()) {
                throw VaultError.InitializationFailed
            }

            // Close existing vault if open
            if (handle != 0L) {
                nativeCloseVault(handle)
                handle = 0L
            }

            val newHandle = nativeCreateVault(path, password)
            if (newHandle == 0L) {
                val error = nativeLastError() ?: "Unknown error"
                throw VaultError.CreationFailed(error)
            }

            handle = newHandle
        }
    }

    /**
     * Open an existing vault
     */
    suspend fun openVault(path: String, password: String) = withContext(Dispatchers.IO) {
        mutex.withLock {
            if (!initialized.get()) {
                throw VaultError.InitializationFailed
            }

            // Close existing vault if open
            if (handle != 0L) {
                nativeCloseVault(handle)
                handle = 0L
            }

            val newHandle = nativeOpenVault(path, password)
            if (newHandle == 0L) {
                val error = nativeLastError() ?: "Unknown error"
                throw VaultError.OpenFailed(error)
            }

            handle = newHandle
        }
    }

    /**
     * Close the current vault
     */
    suspend fun closeVault() = withContext(Dispatchers.IO) {
        mutex.withLock {
            if (handle != 0L) {
                nativeCloseVault(handle)
                handle = 0L
            }
        }
    }

    /**
     * Get vault information
     */
    suspend fun getVaultInfo(): VaultInfo = withContext(Dispatchers.IO) {
        mutex.withLock {
            if (handle == 0L) {
                throw VaultError.InvalidHandle
            }

            val jsonStr = nativeVaultInfo(handle)
                ?: throw VaultError.OperationFailed(nativeLastError() ?: "Unknown error")

            try {
                val json = gson.fromJson(jsonStr, Map::class.java)
                VaultInfo(
                    vaultId = json["vault_id"] as? String ?: "",
                    rootPath = json["root_path"] as? String ?: "",
                    fileCount = (json["file_count"] as? Double)?.toInt() ?: 0,
                    totalSize = (json["total_size"] as? Double)?.toLong() ?: 0L,
                    version = (json["version"] as? Double)?.toInt() ?: 1
                )
            } catch (e: Exception) {
                throw VaultError.JsonParsingFailed
            }
        }
    }

    /**
     * List contents of a directory in the vault
     */
    suspend fun listDirectory(path: String = "/"): List<VaultEntry> = withContext(Dispatchers.IO) {
        mutex.withLock {
            if (handle == 0L) {
                throw VaultError.InvalidHandle
            }

            val jsonStr = nativeVaultList(handle, path)
                ?: throw VaultError.OperationFailed(nativeLastError() ?: "Unknown error")

            try {
                val type = object : TypeToken<List<VaultEntryJson>>() {}.type
                val entries: List<VaultEntryJson> = gson.fromJson(jsonStr, type)
                entries.map { entry ->
                    VaultEntry(
                        name = entry.name,
                        isDirectory = entry.is_directory,
                        size = entry.size
                    )
                }
            } catch (e: Exception) {
                throw VaultError.JsonParsingFailed
            }
        }
    }

    /**
     * Add a file to the vault
     */
    suspend fun addFile(localPath: String, vaultPath: String) = withContext(Dispatchers.IO) {
        mutex.withLock {
            if (handle == 0L) {
                throw VaultError.InvalidHandle
            }

            val result = nativeAddFile(handle, localPath, vaultPath)
            if (result != 0) {
                val error = nativeLastError() ?: "Unknown error"
                throw VaultError.OperationFailed(error)
            }
        }
    }

    /**
     * Extract a file from the vault
     */
    suspend fun extractFile(vaultPath: String, localPath: String) = withContext(Dispatchers.IO) {
        mutex.withLock {
            if (handle == 0L) {
                throw VaultError.InvalidHandle
            }

            val result = nativeExtractFile(handle, vaultPath, localPath)
            if (result != 0) {
                val error = nativeLastError() ?: "Unknown error"
                throw VaultError.OperationFailed(error)
            }
        }
    }

    /**
     * Create a directory in the vault
     */
    suspend fun createDirectory(vaultPath: String) = withContext(Dispatchers.IO) {
        mutex.withLock {
            if (handle == 0L) {
                throw VaultError.InvalidHandle
            }

            val result = nativeMkdir(handle, vaultPath)
            if (result != 0) {
                val error = nativeLastError() ?: "Unknown error"
                throw VaultError.OperationFailed(error)
            }
        }
    }

    /**
     * Remove a file or directory from the vault
     */
    suspend fun removeEntry(vaultPath: String) = withContext(Dispatchers.IO) {
        mutex.withLock {
            if (handle == 0L) {
                throw VaultError.InvalidHandle
            }

            val result = nativeRemove(handle, vaultPath)
            if (result != 0) {
                val error = nativeLastError() ?: "Unknown error"
                throw VaultError.OperationFailed(error)
            }
        }
    }

    /**
     * Change the vault password
     */
    suspend fun changePassword(oldPassword: String, newPassword: String) = withContext(Dispatchers.IO) {
        mutex.withLock {
            if (handle == 0L) {
                throw VaultError.InvalidHandle
            }

            val result = nativeChangePassword(handle, oldPassword, newPassword)
            if (result != 0) {
                val error = nativeLastError() ?: "Unknown error"
                throw VaultError.OperationFailed(error)
            }
        }
    }

    /**
     * Check if a vault is currently open
     */
    val isVaultOpen: Boolean
        get() = handle != 0L
}
